Add TopDownMemoryNodeEliminator

phate opened the pull request:
This PR introduces the TopDownMemoryNodeEliminator. The TopDownMemoryNodeEliminator restricts the number of regions memory states need to be routed through by taking the lifetime of memory nodes into account. For example, the lifetime of a memory node originating from an `alloca` is at most the function the alloca is defined in.

The AgnosticMemoryNodeProvider and RegionAwareMemoryNodeProvider do not take such liftime information into account when producing a memory node provisioning. The idea of the TopDownMemoryNodeEliminator is to take the memory node provisioning produced by these memory node providers as seed provisioning and produce a new provisioning which takes the lifetime information into account. This simple example illustrates the idea:

```
static void
f(int x)
{
  int* alloca;
  *alloca = x;
}
```

The AgnosticMemoryNodeProvider and the RegionAwareMemoryNodeProvider would produce the following provisioning:

RegionEntry: f -> {allocaMemoryNode}
RegionExit: f -> {allocaMemoryNode}

Thus, they would consider the allocaMemoryNode alive at the region entry even though the alloca node was not even defined yet.

The TopDownMemoryNodeEliminator would take this provisioning as input, track the lifetime of alloca nodes, and restrict the provisioning to the following:

RegionEntry: f -> {}
RegionExit: f- > {allocaMemoryNode}

Ultimately, this would mean that the MemoryStateEncoder would need to route no state through the entry of function `f`.

Currently, the TopDownMemoryNodeProvider restricts the lifetime ONLY for alloca nodes and ONLY for region entries.

sjalander commented on jlm/llvm/opt/alias-analyses/MemoryStateEncoder.cpp:
```diff
@@ -390,6 +408,35 @@ class RegionalizedStateMap final {
+  }
+
+  jlm::rvsdg::output&
+  InsertUndefinedMemoryState(jlm::rvsdg::region & region) noexcept
```
Wouldn't it make sense to check for if an undefinedMemoryState already exists?

phate replied:
That is what the `GetOrInsertUndefinedMemoryState` method 20 lines above does, no? It calls this method. Or do I misunderstand your comment?

sjalander replied:
Yes, GetOrInsertUndefinedMemoryState checks if one exists before calling insert, so if this is the only way that insert is called, then the undefined state is only created once.
But, if one instead call insert multiple times, then one would get a new undefined state for each time.
So maybe a JLM_ASSERT could be useful.

phate replied:
`GetUndefinedMemoryState()`, `InsertUndefinedMemoryState()`, and `GetOrInsertUndefinedMemoryState()` are all `private` in the `RegionalizedStateMap` class. `GetUndefinedMemoryState()` and `InsertUndefinedMemoryState()` are only called in `GetOrInsertUndefinedMemoryState()`. Thus, all calls go through `GetOrInsertUndefinedMemoryState()`, which means it is ensured that only one undefined state node per region is created.

phate commented:
@sjalander @haved @caleridas : Could I get a review, please? I would like to merge this one and move on.

haved reviewed (dismissed):
I would like some comments to be added in the places I point out, since I'm still struggling a bit to understand everything going on.

I sort of understand that anything alive inside a function either was alive at the function call, or came alive during the function. The order of traversal visits calls node before the lambda node.

The concept of `AnnotatedLiveNodes` is a bit weird to me, as it seems having a single direct call makes `EliminateTopDownLambdaEntry` assume the function can't be called indirectly?

I'm also a bit unsure how malloc memory objects become "alive", other than being included from the seedProvisioning. The only places that read the seed provisioning without also intersecting it with HashSets from the `Context_` are `EliminateTopDownLambdaEntry` when the lambda is not annotated, and `InitializeLiveNodesLambda` when the lambda is a tail node. What about all other lambdas? Is the idea that any lambda that is not a tail node, must be called from somewhere, and all malloc memory objects will be alive at that call site?

It seems `InitializeLiveNodes` handles functions escaping the module, which can be called from external call sites, thus making it hard to eliminate live memory objects.
However, it is only marking tail nodes, which ignores any nodes that are exported if they are also depended on by some other node.

When trying to write these comments, I'm discovering that any issue I can think of seems to be covered when I really dig into it, but I can't guarantee that I've covered every possible weird program.

phate commented:
Thanks for the comments. They made me realize that there is still some more work to do.

phate commented on jlm/llvm/opt/alias-analyses/TopDownMemoryNodeEliminator.cpp:
```diff
@@ -0,0 +1,971 @@
+      auto & lambdaNode = *callTypeClassifier->GetLambdaOutput().node();
+      return GetLambdaEntryNodes(lambdaNode);
+    }
+    else if (callTypeClassifier->IsRecursiveDirectCall())
```
Utilized a `||`

phate commented on jlm/llvm/opt/alias-analyses/TopDownMemoryNodeEliminator.cpp:
```diff
@@ -0,0 +1,971 @@
+      auto & lambdaNode = *callTypeClassifier->GetLambdaOutput().node();
+      return GetLambdaExitNodes(lambdaNode);
+    }
+    else if (callTypeClassifier->IsRecursiveDirectCall())
```
Utilized `||`

phate commented on jlm/llvm/opt/alias-analyses/TopDownMemoryNodeEliminator.cpp:
```diff
@@ -0,0 +1,973 @@
+  {
+    if (!HasLiveNodesSet(region))
+    {
+      LiveNodes_[&region] = {};
```
I left the code as it is for clarity.

phate commented on jlm/llvm/opt/alias-analyses/TopDownMemoryNodeEliminator.cpp:
```diff
@@ -0,0 +1,973 @@
+void
+TopDownMemoryNodeEliminator::EliminateTopDownRootRegion(rvsdg::region & region)
+{
+  rvsdg::bottomup_traverser traverser(&region);
```
I added comments to clarify the situation

phate commented on jlm/llvm/opt/alias-analyses/TopDownMemoryNodeEliminator.cpp:
```diff
@@ -0,0 +1,971 @@
+  auto & provisioning = Context_->GetProvisioning();
+  auto & seedProvisioning = Context_->GetSeedProvisioning();
+
+  if (Context_->HasAnnotatedLiveNodes(lambdaNode))
```
I added more comments to clarify.

phate commented on jlm/llvm/opt/alias-analyses/TopDownMemoryNodeEliminator.cpp:
```diff
@@ -0,0 +1,973 @@
+
+  EliminateTopDownRootRegion(phiSubregion);
+  computeLiveNodes(phiSubregion);
+  EliminateTopDownRootRegion(phiSubregion);
```
I would like to limit the number of live nodes to those that are only considered live. The live nodes in the `seedProvisioning` might have superfluous live nodes. That is why the call to `EliminateTopDownRootRegion` is done twice:

1. Compute live nodes for each lambda in the phi region by calling `EliminateTopDownRootRegion`
2. Unify the live nodes for all lambdas
3. Ensure all live nodes to the region entry/exits are properly annotated by calling `EliminateTopDownRootRegion` again.

I added some more comments. I hope this clarifies it.

phate commented on jlm/llvm/opt/alias-analyses/TopDownMemoryNodeEliminator.cpp:
```diff
@@ -0,0 +1,971 @@
+    JLM_ASSERT(is<PointerType>(output.type()));
+    auto & registerNode = PointsToGraph_.GetRegisterNode(output);
+
+    util::HashSet<const PointsToGraph::MemoryNode*> memoryNodes;
```
That might be, but I have not really thought out this entire infrastructure (yet). I would like to keep it where it is right now and solve this later. I have plans to get rid off this entire `Provisioning` classes and move these annotations in the `PointsToGraph`.

phate commented on jlm/llvm/opt/alias-analyses/TopDownMemoryNodeEliminator.cpp:
```diff
@@ -0,0 +1,973 @@
+  }
+
+  [[nodiscard]] const util::HashSet<const PointsToGraph::MemoryNode *> &
+  GetCallEntryNodes(const CallNode & callNode) const override
```
You are absolutely right, there are is documentation on these methods. However, the documentation should happen in `MemoryNodeProvisioning` class as most of these methods are just overrides from this class. I would prefer to do this in another PR.

phate commented:
I made the following changes:
1. Sprinkled comments over the code to (hopefully) improve the understanding.
2. Added some more `JLM_ASSERT`s to also aid understanding.
3. Fixed a bug that dawned on me after Håvard's comments (removing all allocas instead of just unescaped ones)
4. Addressed all the comments

I hope that this will improve the understanding of this PR.

haved reviewed (approved):
Looks great, the comments helped a lot. It now makes sense why inter-functional nodes are processed bottom up (calls go up), while intra-functional nodes are processed top down (an alloca is only alive if its output can reach a given region, which has to be further down the DAG).

phate commented on jlm/llvm/opt/alias-analyses/TopDownMemoryNodeEliminator.cpp:
```diff
@@ -0,0 +1,1003 @@
+/*
+ * Copyright 2023 Nico Reißmann <nico.reissmann@gmail.com>
```
I would need another approval for it. Not worth it. This is currently broken in the code base anyway. We would need something that checks this at PR time to get this right.
